internal/rbd/rbd_journal.go

```diff
@@ -388,6 +388,18 @@ func (ri *rbdImage) repairImageID(ctx context.Context, j *journal.Connection, fo
 func reserveSnap(ctx context.Context, rbdSnap *rbdSnapshot, rbdVol *rbdVolume, cr *util.Credentials) error {
 	var err error
 
+	// restore original values in case of an error
+	origReservedID := rbdSnap.ReservedID
+	origRbdSnapName := rbdSnap.RbdSnapName
+	origVolID := rbdSnap.VolID
+	defer func() {
+		if err != nil {
+			rbdSnap.ReservedID = origReservedID
+			rbdSnap.RbdSnapName = origRbdSnapName
+			rbdSnap.VolID = origVolID
+		}
+	}()
```

Comment:
@nixpanic any specific reason to preserve and restore these values? i assume if the reservation is done properly these values will be set and not overridden to any values isnt it?

Reply:
there are these two calls that overwrite the attributes in any case (success of error):
```
rbdSnap.ReservedID, rbdSnap.RbdSnapName, err = j.ReserveName(...
```
and
```
rbdSnap.VolID, err = util.GenerateVolID(...
```
So, if something goes wrong, it looks nicer to reset the attributes that may have been modified by a failing call.

Reply:
Setting those attributes only at the end of the function is tricky too, `undoSnapReservation()` expects them to be set in order to clean things up in the journal.

Reply:
@Madhu-1 if they are not expected to be set, shall I add a check and error message in case they are not an empty string?

Reply:
currently we dont have any case where they are set before calling this function. there are callers doing undo reservation and that will fail or not required

* Call Reserve to do reservation
* Undo if there is any error

we updated the current code to do undo internally if something fails.

Doesnt hurt to have this code but this will result in extra error/logs when we try to call the Undo function multiple times.


Let me know what do you prefer

1. Merge the current changes
2. Add a check to ensure the values are not set before calling `reserveSnap` internal function
3. Dont revert back and allow override.

Reply:
The current PR as it is shows what a function should do/restore. There are no callers that will call `undoSnapReservation()` in case `reserveSnap()` fails, so there should not be additional logs for that.

When a function returns an error, there should be no need for the caller to undo whatever internals the function may, or may not, have done. Everything internal to the function should stay there. For the more complex cases, a special `undo` function can be returned, similar to what [`Manager.getGroupUUID()`](https://github.com/ceph/ceph-csi/blob/cea8bf8110531284d272582b9c536ffdb2ece82b/internal/rbd/manager.go#L117-L166) does.

Merging the current PR as it is should be fine IMHO, @Madhu-1, but I am also fine with improving it some way you recommend.

Reply:
No, it would not. In case there is an error returned by `reserveSnap()`, the `defer` where `undoSnapReservation()` is called will never be reached.